model_struct.go

```diff
@@ -17,6 +17,8 @@ var DefaultTableNameHandler = func(db *DB, defaultTableName string) string {
 	return defaultTableName
 }
 
+var mutex sync.Mutex
```

Comment:
Why you added this to the package and not into ModelStruct type?

Reply:
@mtfelian doesn't this depend on what context the go routines are called in? If I want a global mutex but the scope is inside the struct it might not actually work as intended. 

That said, I am not familiar with this code base and used the 'nuclear' option to be safe. I was thinking about this PR as a discussion point more than an actual PR. The state of the linked bug was "can't reproduce" but the issue clearly exists. This PR fixes the bug, but there are other concerns I have with this PR in general: how this might impact performance and if there are any deadlock scenarios. 

Thanks!

Reply:
maybe @jinzhu can chime in here

Reply:
@jontonsoup think this should not have a visible affect on performance due to no "hard" methods are inside. To make it faster, you can set/get the `defaultTableName` field with mutex. 

I tested my fork (i hurried to make a PR, but the Mutex there is inside the struct) with our production code with transactions, multithreading and other things and it fixes the race. 

So i think you can move it inside the struct: 

```
// ModelStruct model definition
type ModelStruct struct {
	PrimaryFields    []*StructField
	StructFields     []*StructField
	ModelType        reflect.Type
	defaultTableName string

	l sync.Mutex
}
```

Reply:
I think we could just use sync.Map to avoid the race condition

Reply:
@jinzhu i created another PR https://github.com/jinzhu/gorm/pull/2022 where it's changed to sync.Map